<?xml version="1.0" ?>

<robot name="axis_camera">

  <!-- LINKS -->

  <link name="wall_link">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>

  <link name="pan_link">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>

  <link name="tilt_link">
    <visual>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
      <origin xyz="0 0 0"/>
    </visual>

    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>

  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.005 0.03 0.01"/>
      </geometry>
      <origin xyz="0 0 0"/>
    </visual>

    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>

  <!-- JOINTS -->
  <joint name="wall_to_pan" type="revolute">
    <axis xyz="0 0 1"/>
    <origin xyz="0.0 0.0 0.0"/>
    <parent link="wall_link"/>
    <child link="pan_link"/>
    <limit effort="100000" lower="-600.28" upper="600.28" velocity="30"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>

  <joint name="pan_to_tilt" type="revolute">
    <axis xyz="0 1 0"/>
    <origin xyz="0.0 0.0 0.0"/>
    <parent link="pan_link"/>
    <child link="tilt_link"/>
    <limit effort="100000" lower="-100.57" upper="100.57" velocity="30"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>

  <joint name="tilt_to_camera" type="fixed">
    <origin rpy="-1.5707963 0 -1.5707963" xyz="0.05 0 0"/>
    <parent link="tilt_link"/>
    <child link="camera_link"/>
    <limit effort="0.0" lower="0.0" upper="0.0" velocity="0.0"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>

  <!-- PLUGINS -->
  <gazebo reference="camera_link">
    <gravity>0</gravity>
  </gazebo>

  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="wall_to_pan">
      <command_interface name="velocity">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="pan_to_tilt">
      <command_interface name="velocity">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>$(find axis_camera)/config/axis_camera_controller.yaml</parameters>
    </plugin>
  </gazebo>

  <gazebo reference="camera_link">
    <sensor name="axis_camera" type="camera">  <!-- type="camera" -->
      <update_rate>50</update_rate>

      <camera>
        <horizontal_fov>1.078613</horizontal_fov>
        <vertical_fov>0.649262</vertical_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.05</near>
          <far>300</far>
        </clip>
      </camera>

      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <!-- ONLY TAG WORKING -->
        <!-- <ros>
          <namespace>custom_ns</namespace>
        </ros> -->

        <!-- TAGS NOT WORKING -->
        <!-- <update_rate>0</update_rate>
        <baseline>0.2</baseline>
        <always_on>0</always_on>
        <robotNamespace>single_rrbot</robotNamespace>
        <cameraName>axis_cameraaaaaaaa</cameraName>
        <imageTopicName>immmmmage_raw</imageTopicName>
        <cameraInfoTopicName>cccccamera_info</cameraInfoTopicName>
        <depthImageTopicName>iiiiimage_raw</depthImageTopicName>
        <depthImageInfoTopicName>/deeeeepth/camera_info</depthImageInfoTopicName>
        <pointCloudTopicName>/depppppth/points</pointCloudTopicName>
        <frameName>camera_link</frameName>
        <pointCloudCutoff>0.5</pointCloudCutoff>
        <pointCloudCutoffMax>3.0</pointCloudCutoffMax>
        <distortionK1>0.00000001</distortionK1>
        <distortionK2>0.00000001</distortionK2>
        <distortionK3>0.00000001</distortionK3>
        <distortionT1>0.00000001</distortionT1>
        <distortionT2>0.00000001</distortionT2>
        <CxPrime>0</CxPrime>
        <Cx>0</Cx>
        <Cy>0</Cy>
        <focalLength>0</focalLength>
        <hackBaseline>0</hackBaseline> -->
      </plugin>
    </sensor>
  </gazebo>

</robot>
